--- 
title: "The Epidemiology of _P. vivax_ Among Adults in the Democratic Republic of the Congo" 
author: "Nicholas F. Brazeau" 
date: "`r format(Sys.time(), '%B %d, %Y')`" 
output: 
  html_document: 
    highlight: tango 
    theme: lumen 
    toc: yes 
    toc_float: yes 
    toc_depth: 3 
    code_folding: hide
editor_options:  
  chunk_output_type: console
--- 

<!-- <style> -->
<!--   .col2 { -->
<!--     columns: 2 200px;         /* number of columns and width in pixels*/ -->
<!--     -webkit-columns: 2 200px; /* chrome, safari */ -->
<!--     -moz-columns: 2 200px;    /* firefox */ -->
<!--   } -->
<!-- <style> -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, warning = F, message = F)
```
```{r, imports}
source("~/Documents/GitHub/VivID_Epi/R/00-functions_basic.R")
library(tidyverse)
library(DT)
library(kableExtra)
# special plots
library(ggdag)
library(ggcorrplot)

options(scipen=999)
```

## Overview
Here we have screened over 17,000 adults in the DRC to determine the prevalence and risk factors associated with _P. vivax_ infection. 

### Main Question & Approaches
1. Who is being infected with _P. vivax_ (traditional Epi approach)    
    * Bivariate ORs (GEE); Table 1 and Table 2 combined (keeping in mind feedback from MDF's reviewer)  
1. Who is being infected with _P. vivax_ and **Where** are Infections Occurring ("Prediction Models"; Less Traditional Epi)
    * Bayesian Hierarchial/Spatial Prediction Model (PrevMap, INLA, BRMS)
1. Where did these _P. vivax_ Infections Originate 
    * _PvAMA1_ amplicon 
1. Are antimalarial interventions equally effective against _P. vivax_ and _P. falciparum_     
    * Unconfounded ORs (Inverse Probability Weights using an Ensemble Supervised ML "Stack")
    * This will allow us to make causal statements such as: "Although bed-nets prevented ## cases of _P. falciparum_, they only prevented ## cases of _P. vivax_."
    
### Methodological Considerations
DHS survey weights are included in the risk factor effect estimates using GEE (per DHS's recommendation). 

### Background
Extensive recoding was performed both for merging samples to the CD2013 DHS datasets and for extracting covariates from several publicly available resources (IUCN, OSM, etc). Descriptions are within the _Google Docs_ below. 
  
1. [Sample Selection](https://docs.google.com/document/d/1qGSKnlr5RjYjz2cURATWw4dgL58u561mZfq9qDpidWk/edit?usp=sharing)  
1. [Sample Screening]()   
1. [Covariate Recoding](https://drive.google.com/file/d/1g09QSawu2z8sqe6HOFo1DNoQGt17Bcy1/view?usp=sharing)   
1. [_P. vivax_ Whole Genomes](https://drive.google.com/file/d/1_NU9nQHEkw22qfcXuVUcq65H2Wm-3G0V/view?usp=sharing) -- Curated and downloaded by NFB & APM



## Risk Factors -- Traditional Epi 
See "Covariate Recoding" and "Covariate Association Predictions" above for a complete list of predicted asscociations. Of note, among the ~500 potential infections, only three individuals were Duffy-positive (heterozygtoe genotype). Given this near fixation of the Duffy negative phenotype, it was not considered further in risk factor analyses. 
```{r import risk factors}

load("~/Documents/GitHub/VivID_Epi/results/bivariate_model_results.rda")

```

### Distributions & Bivariate Estimates {.tabset .tabset-fade .tabset-pill}
All continuous variables are scaled for parametric estimates.   Signficant variables are highlighted in blue and all _cluster-level covariates are italicized_. 
  

#### _P. vivax_ Risk Factors 
```{r}
pvivriskfactortable <- pvivriskfactortable %>% 
  dplyr::mutate(levels = ifelse(
    stringr::str_detect(Covariates,
                    "\\s\\s\\s[a-z]|\\s\\s\\s[A-Z]"), Covariates, ""),
    Covariates = ifelse(
       stringr::str_detect(Covariates,
                    "\\s\\s\\s[a-z]|\\s\\s\\s[A-Z]"), "", Covariates)
  ) %>% 
  dplyr::select(c("var_label", "Covariates", "levels", "Pvivax-Negative", "Pvivax-Positive", "estimate", "conf.low", "conf.high"))

sigrows <- which(pvivriskfactortable$conf.low < 1 & pvivriskfactortable$conf.high < 1 
                 | pvivriskfactortable$conf.low > 1 & pvivriskfactortable$conf.high > 1)

clusterrows <- which(grepl("_clst", pvivriskfactortable$Covariates))

pvivriskfactortable[is.na(pvivriskfactortable)] <- ""
pvivriskfactortable %>% 
  knitr::kable(., "html") %>%  # note must tell kableExtra html or latex
  kableExtra::kable_styling(bootstrap_options = "striped", full_width = F, position = "center") %>% 
  kableExtra::row_spec(row = clusterrows, italic = T) %>% 
  kableExtra::row_spec(row = sigrows, bold = T, background = "#deebf7")

```

#### _P. falciparum_ Risk Factors 
```{r}

pfalriskfactortable <- pfalriskfactortable %>% 
  dplyr::mutate(levels = ifelse(
     stringr::str_detect(Covariates,
          "\\s\\s\\s[a-z]|\\s\\s\\s[A-Z]"), Covariates, ""),
    Covariates = ifelse(
       stringr::str_detect(Covariates,
            "\\s\\s\\s[a-z]|\\s\\s\\s[A-Z]"), "", Covariates)
  ) %>% 
  dplyr::select(c("var_label", "Covariates", "levels", "Pfalciparum-Negative", "Pfalciparum-Positive", "estimate", "conf.low", "conf.high"))

sigrows <- which(pfalriskfactortable$conf.low < 1 & pfalriskfactortable$conf.high < 1 
                 | pfalriskfactortable$conf.low > 1 & pfalriskfactortable$conf.high > 1)

clusterrows <- which(grepl("_clst", pfalriskfactortable$Covariates))

pfalriskfactortable[is.na(pfalriskfactortable)] <- ""
pfalriskfactortable %>% 
  knitr::kable(., "html") %>%  # note must tell kableExtra html or latex
  kableExtra::kable_styling(bootstrap_options = "striped", full_width = F, position = "center") %>% 
  kableExtra::row_spec(row = clusterrows, italic = T) %>% 
  kableExtra::row_spec(row = sigrows, bold = T, background = "#deebf7")

```


    
## Bayesian Heirarchial/Spatial Prediction Models

### Weighted Prevalence {.tabset .tabset-fade .tabset-pill}
Province and cluster point estimates were calculated directly from the CD2013 sampling weights. The raster surfaces are Smooth Guassian Processes from `PrevMap` using a MLE approach. Moran's I was calculated from greater circle distance between clusters.

```{r}
prevmaprasterplots <- readRDS("~/Documents/GitHub/VivID_Epi/results/prevmap_raster_plots.rds")
load("~/Documents/GitHub/VivID_Epi/results/basic_maps_results.rda")

```

#### _P. vivax_

```{r, fig.width=11, fig.height=8, fig.align="center"}
prevmaprasterplots$moranI[[2]]

gridExtra::grid.arrange(
                        pntestmaps[[2]], 
                        caseprevmaps[[2]], 
                        prevmaprasterplots[[2]],
                        ncol=3, top=grid::textGrob("P. vivax Prevalence in CD2013 DHS", gp=grid::gpar(fontsize=15, fontfamily = "Arial", fontface = "bold"))) 

```

#### _P. falciparum_
```{r, fig.width=11, fig.height=8, fig.align="center"}
prevmaprasterplots$moranI[[1]]

gridExtra::grid.arrange(
                        pntestmaps[[1]], 
                        caseprevmaps[[1]], 
                        prevmaprasterplots[[1]],
                        ncol=3, top=grid::textGrob("P. falciparum Prevalence in CD2013 DHS", gp=grid::gpar(fontsize=15, fontfamily = "Arial", fontface = "bold"))) 


```

#### _P. ovale_
```{r, fig.width=11, fig.height=8, fig.align="center"}
prevmaprasterplots$moranI[[3]]

gridExtra::grid.arrange(
                        pntestmaps[[3]], 
                        caseprevmaps[[3]], 
                        prevmaprasterplots[[3]],
                        ncol=3, top=grid::textGrob("P. ovale Prevalence in CD2013 DHS", gp=grid::gpar(fontsize=15, fontfamily = "Arial", fontface = "bold"))) 


```

###

#### Ape Habitat Overlap
```{r}
plot( aperange_nhapv )
```

## _P. vivax_ Phylogeographics 
```{r}
# load("~/Documents/GitHub/VivID_Epi/WetLabWork/PvampliconDesign/potential_primer_variants.rda")
```
Based on the ~735 samples that Andrew and I found and previous estimates of $F_{st}$ by Hupalo et al. 2016 [PMID: 27348298](https://www.ncbi.nlm.nih.gov/pubmed/27348298) (see Supplementary Table 2 & Figure 5), I identified small regions in several genes that was "amplifi-able" for our low-quality DNA (i.e. ~300 bp) and were phylogeographically informative. Of those candidates, _PvAMA1_ proved to be best under our laboratory conditions.   

### Phylogenetic Tree of _PvAMA1_
  * No DRC samples yet but NFB randomly selected 96*2 and they are plated, ready to PCR'ed/sequenced.   
  * Will need to subset to monoclonal among 735 publicly available.  
  
```{r}
ggplot() +
  annotate("text", x=0, y=0, label= "Wet Lab Work in Progress", color = "red") +
  vivid_theme +
  theme(axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())
```



## Antimalarial Interventions
Based on my [DAG](http://dagitty.net/development/dags.html?id=jKNQhB#), I have identifed potential confounders of the _Plasmodium_ - Intervention relationship. We will use Invese Probability Confounding Weights to remove the bias due to confounding and Inverse Probability Sampling Weights to account for the DHS sampling structure (per DHS recommendations) in order to arrive at a **causal estimate**. The interventions that will be compared between _P. vivax_ and _P. falciparum_ are:   
1. Bednets (LLIN)  
1. Housing Materials  
1. Healthcare Insurance/Access  
  
These interventions were all identified in the literature as recent, scalable, cost-effective measures for combatting malaria. 


#### Due Diligence
```{r import due diligence}

dtrskfctr.corr <- readRDS("~/Documents/GitHub/VivID_Epi/data/derived_data/covariate_correlation_data.rds")

```

##### Model Multicollinearity 
```{r, fig.width = 8, fig.height = 8 }
ggcorrplot::ggcorrplot(dtrskfctr.corr, 
                       type = "lower",
                       outline.color = "white",
                       ggtheme = ggplot2::theme_gray,
                       colors = c("#313695", "#ffffbf", "#a50026"),
                       hc.order = F) +
  ggtitle("Covariate Correlation Matrix") +
  theme(plot.title = element_text(size = 14, family = "Arial", face = "bold", vjust = 0.5, hjust = 0.5),
        legend.position = "right", 
        legend.text = element_text(angle = 0),
        axis.text = element_text(size = 5),
        panel.border = element_blank())

```

##### Missingness
**For now, have gone to a complete case analysis (MCAR assumption). In the future, I should consider multiple imputation because a few of the cluster did not have antimalarial-drug use, which means there is a MAR assumption at the very least**.
```{r}
# mice::md.pattern(llin, rotate.names = T)

```

##### Causal Assumptions
1. Exchangeability -- No unmeasured confounding   
    * Going to include all measured confounders and not just minimally sufficient to help with this. 
1. Positivity 
    * IPW implicitly assumes that all individuals have a non-zero probability of receiving either exposure/treatment (i.e. must have an individual through all cells).  
1. Consistency 
    * Cannot guarantee but should be OK. 
1. Temporality  
    * Cross sectional, so cannot guarantee
1. No Interference  
    * Infectious disease, so certainly cannot guarantee but have attempted to somewhat control for this dependence by accounting for physical space 
    
    
### IPTW Effect Esimates {.tabset .tabset-fade .tabset-pill}

```{r}
load("~/Documents/GitHub/VivID_Epi/results/llin_iptw_models.rda")

# load
# load
# lapply() %>% bindrows -- and then filter below

```

#### LLIN

<!-- <div class="col2"> -->
##### Unweighted "Table One"
```{r}
tableone::kableone(llintable_orig)
```

##### IPTW-weighted "Table One"
```{r}
tableone::kableone(llintable_iptw)
```
<!-- </div> -->

```{r}

pan.llin.cest.tidy %>% 
  knitr::kable(., "html") %>% 
  kableExtra::kable_styling(bootstrap_options = "striped", full_width = F, position = "center")

pan.llin.cest.tidy %>% 
  dplyr::mutate(y = 1:nrow(.)) %>% 
  ggplot(data = .) +
  geom_vline(aes(xintercept = 1), size = 1.5, color = "#AA0A3C", alpha = 0.9, linetype='dashed') +
  geom_point(aes(x=OR, y = y, color = factor(species)), size = 3) +
  geom_errorbarh(aes(xmin = L95, xmax = U95, y = y, color = factor(species)), size = 1.5) +
  scale_color_manual("Species", values = c("#00A0FA", "#006E82")) +
  ggtitle(expression(paste(italic("P. vivax"), " and ", italic("P. falciparum"), "Intervention Causal Estimates"))) +
  xlab("Odds Ratio") +
  vivid_theme + 
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.line.x = element_line(colour = "#000000", size = 1, linetype = "solid"),
        legend.position = "right",
        legend.text = element_text(face = "italic", angle = 0))

  

```

#### Housing Materials
```{r}


```
    
#### Health Insurance 
```{r}


```


## Apendix
### Derivation of IPTWs

Assume that we have an treatment (A) and an outcome (Y) with confounding variables Z1 and Z2. The confounding variables will be abbreviated **Z** below. 
```{r, fig.align='center', fig.width=5, fig.height=4}
dagify(Y ~ A,
       A ~ z1,
       Y ~ z1,
       A ~ z2,
       Y ~ z2) %>% 
  ggdag() + 
  theme(legend.position = "none",
        panel.background = element_blank(),
        panel.grid = element_blank())
```

### _Some basic reminders_: 
**Joint Probabilities**   
$$P(Y,X) = P(Y|X)P(X)$$
$$P(Y,X) = P(X,Y)$$ 
**Law of Total Probability**   
$$P(Y) = \sum P(Y|X)P(X)$$    

**Causal Inference**  
Assuming $Y$ is a dichotomized outcome, then $E(Y^a) = P(Y^a)$. Note, this I am dropping time aspect from classic equation of $F_{Y^a}(t) = P(Y^a \leq t)$. Moreover, assume that $Y^a \bot A | Z$, exchangeability, and counterfactual consistency, then:

$$P(Y | A = a) = \sum_Z P(Y|A=a, \mathbf Z)P(\mathbf Z)$$

Therefore, we just have to reduce the confounders by creating weights that account for the probability that we received treatment:    

$$W^A = \frac{1}{g[A|\mathbf Z]}$$
Therefore, our new weighted outcome is $E(Y^a) = P(Y^a) = E[\frac{I(A=a)Y}{(W^A)^{-1}}]$.

$$E(Y^a) = \sum_Z \frac{P(Y, A, \mathbf Z)}{(W^A)^{-1}}$$
$$E(Y^a) = \sum_Z \frac{P(Y|A)P(A|Z)P(Z)}{P(A|Z)}$$








