---
title: "Find Fasta Sequences for Primer3 based on Potential Primers"
author: "Nicholas F. Brazeau"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document:
    highlight: textmate
    theme: lumen
    toc: yes
    toc_float: yes
    toc_depth: 2
    code_folding: hide
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F, fig.width = 11, fig.height = 8)
#.................................
# imports
#.................................
library(tidyverse)
devtools::install_github("IDEELResearch/vcfRmanip")
library(vcfRmanip)
library(vcfR)
library(Biostrings)
```

# Update
Updated on April 25, 2019 to design an outer primer set (nested pcr approach) for _Pvama1_, as it showed some promise on initial run ( _Pvs47_ failed) and _pvmsp1_ is already a nested primer design. 

# Overview
Overall, uninspiring results. Will design for: 

  1. _pvs47_   
  2. _pvama1_   
  3. _pvmps1_  


```{r}

load("~/Documents/GitHub/VivID_Epi/WetLabWork/PvampliconDesign/potential_primer_variants.rda")

# subset to ama1 and pvs47

bd <- bd %>% 
  dplyr::filter(geneid %in% c("PVP01_1208000", "PVP01_0934200"))


# read fasta
pvp01 <- Biostrings::readDNAStringSet("~/Documents/MountPoints/mountIDEEL/resources/genomes/Pvivax/genomes/PvP01.fasta", "fasta")
chr12 <- pvp01[ pvp01@ranges@NAMES == "PvP01_12_v1" ]  # for pvs47
chr09 <- pvp01[ pvp01@ranges@NAMES == "PvP01_09_v1" ]  # for ama1
chr07 <- pvp01[ pvp01@ranges@NAMES == "PvP01_07_v1" ]  # for msp1
# pvmsp1 already designed


```

```{r}

bd$pos <- map(bd$vcfRobj, function(x){ 
  ret <- vcfR::getFIX(x)
  ret <- ret[,2]
  return(ret)
})


```

### _pvs47_ Design

From `01-Potential_Primer_interrogate.Rmd` report, I would like to capture the SNPs at the positions: `r paste("323549", "323599", "323603", "323611", "323612", "323718", "323761", "323777")`. Al together this is a 229 base pair region. However, the position 323549 is very close to the start of the gene (_starts at 323534_ which is 15 base pairs upstream). **This means I am going to have start my forward primer in a 5' UTR**. Going to have that as the lower limit and will only go 50 base-pairs up. With 20 base-pair primer expected, this will give me 

#### Conditions on `Primer-Blast`

Used the sequence below. Set the min PCR product size to 300 and the max 600. Then set organism "Plasmodium vivax (taxid:5855)". All other parameters were left at default. The "Primer Set 4" was selected as it had 55% GC content and then the lowest _Self complementarity_ score coupled with a slightly higher melt temperature than "Primer Set 3" and had a **300 bp** target (while still getting our SNPs.)

```{r}

# note these are acting as 1-based which is the same as the vcf


pvs47 <- toString(subseq(chr12, (323549-100), (323777+100)))

# forward
vmatchPattern("GACCACTAAACGCGAAGTGC", chr12)

# reverse
vmatchPattern(reverseComplement(DNAString("TTGGTGAAGCAGTTGGCTGG")), chr12)


```



### _pvama_ Design

From `01-Potential_Primer_interrogate.Rmd`, I would like to capture the SNPs at the positions: `r paste("1460429", "1460433", "1460434", "1460440", "1460441", "1460450", "1460494", "1460608", "1460628", "1460662")`. All together this is a 233 base pair region. Note, end of gene is 1460945, so have some room. 

```{r}

ama1 <- toString(subseq(chr09, (1460429-100), (1460628+100)))

vmatchPattern("ATACCAAACCGACTTGCCTCA", chr09)
vmatchPattern(reverseComplement(DNAString("TTTCCGCCCTTTTCTCTACACA")), chr09)
```

Get all the SNPs with these and don't hit variation that we have in our VCF (one site with one het call in ~800 samples) that is missing from blast. This is important if we want the primers to land. GC content only 45% but that should be OK. With all this in mind "Primer Set 1" was selected.



### _pvama_ Nested Design
Now going to design outer primers for a nested approach. Above, we know that our forward and reverse primers (eg set 1 from above) fall here, respectively: 
```{r}
prmrs <- rbind( data.frame( vmatchPattern("ATACCAAACCGACTTGCCTCA", chr09)[[1]] ),
         data.frame( vmatchPattern(reverseComplement(DNAString("TTTCCGCCCTTTTCTCTACACA")), chr09)[[1]] )
  ) %>% 
    cbind(Primer = c("Forward", "Reverse"), .)

prmrs %>% 
  DT::datatable()

```
  
Now going to leverage that information for outer primers. Going to expand out another 250 bp from the ends of the primers and ask for `Primer-blast` to find a proce with the settings: min PCR product size to 400 and the max 600. Then set organism "Plasmodium vivax (taxid:5855)". All other parameters were left at default.
```{r}

ama1.outer <- toString(subseq(chr09, 
                              (prmrs$start[prmrs$Primer == "Forward"]-250), 
                              (prmrs$end[prmrs$Primer == "Reverse"]+250)))


```

`Primer-Blast` asked to match on this *"XM_001615397.1	Plasmodium vivax apical merozoite antigen 1 (PVX_092275), partial mRNA"* .... which is a Sal1 version...which should be fine (annoying default behavior). Will check that they match and work below. The third primer set fit the bill with falling approximately 100-bp upstream and downstream of each of the inner primers designed above. Plus GC content is 55% and the reverse has a GC hammer at beginning. 


```{r}


vmatchPattern("AGTGCCTTTCTTCCAGTGGG", chr09)
vmatchPattern(reverseComplement(DNAString("GCACACGCCACCAGTTATTC")), chr09)


prmrs.all <- cbind(status = "inner", prmrs) %>% 
  rbind.data.frame(
    rbind( data.frame( vmatchPattern("AGTGCCTTTCTTCCAGTGGG", chr09)[[1]] ),
           data.frame( vmatchPattern(reverseComplement(DNAString("GCACACGCCACCAGTTATTC")), chr09)[[1]]
                       )
           ) %>% 
      cbind(status = "outer", Primer = c("Forward", "Reverse"), .)
    )

prmrs.all %>% 
  DT::datatable()


```


### _pvmsp1_ 

Where do they land?
```{r}
# inner primers published by JESS
vmatchPattern("GAGCTCTACAAGACGCACTT", chr07)
vmatchPattern(reverseComplement(DNAString("CTCCARCTCGGCCTTTTT")), chr07, max.mismatch = 1)


```

### _pv18s snounou_ 

Where do they land?
```{r}
# snounou primers
snounouhitF <-vmatchPattern(reverseComplement(DNAString( "CGCTTCTAGCTTAATCCACATAACTGATAC" )), pvp01[ !grepl("Transfer", pvp01@ranges@NAMES) ] )
snounouhitR <- vmatchPattern("ACTTCCAAGCCGAAGCAAAGAAAGTCCTTA", pvp01[ !grepl("Transfer", pvp01@ranges@NAMES) ])

# apparently it is only hitting on CHR2 even though it is a multicopy gene?
vmatchPattern(reverseComplement(DNAString( "CGCTTCTAGCTTAATCCACATAACTGATAC" )), pvp01[  pvp01@ranges@NAMES == "PvP01_02_v1" ] )
vmatchPattern("ACTTCCAAGCCGAAGCAAAGAAAGTCCTTA",  pvp01[  pvp01@ranges@NAMES == "PvP01_02_v1" ] )



```


