---
title: "Primer Query for Population Structure"
author: "Nicholas F. Brazeau"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document:
    highlight: textmate
    theme: lumen
    toc: yes
    toc_float: yes
    toc_depth: 2
    code_folding: hide
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F, fig.width = 11, fig.height = 8)
#.................................
# imports
#.................................
library(tidyverse)
devtools::install_github("IDEELResearch/vcfRmanip")
library(vcfRmanip)
library(vcfR)
```

# Overview
Is there diversity in pv18s relative to the "vivax-like" sequences?  
  
First, let's compare to the _Pv_ global vcf that APM and I made. Next let's just blast the snounou sequence. Wary of the fact that there are several other "malaria"-like sequences out there (hepatocytosis, etc.)


```{r, results='hide'}



#.................................
# read in
#.................................
# note, I subsetted this vcf on the command line previously
dv <- vcfR::read.vcfR(file = "~/Documents/MountPoints/mountedMeshnick/Projects/VivID_Epi/WetLabWork/PvampliconDesign/vcf/pv18s_cmpr.VQSR900.vcf.gz")
bd <- readr::read_tsv(file = "~/Documents/GitHub/VivID_Epi/WetLabWork/PvampliconDesign/primerbed_fastas/pv18s.bed")
mt <- readxl::read_excel(path = "~/Documents/GitHub/VivID_Epi/WetLabWork/PvampliconDesign/curated_samples_20181211.xlsx")

#.................................
# datawrangle
#.................................
bd <- bd %>%
  dplyr::rename(seqname = chr)
bdlist <- split(bd, factor(1:nrow(bd)))

bd$vcfRobj <- purrr::map(bdlist, vcfRmanip::vcfR2SubsetChromPos, vcfRobject = dv)
bd$nvar <- unlist(purrr::map(bd$vcfRobj, function(x){ return(nrow(x@gt)) }))

#.................................
# data manip for plotting
#.................................

bd$plotdf <- purrr::map(bd$vcfRobj, function(x, metadata = mt){
                              ret <- cbind.data.frame(CHROM = vcfR::getCHROM(x),
                                     POS = vcfR::getPOS(x),
                                     vcfR::extract.gt(x, element = "GT")) %>% 
                                     gather(., key = "iid", value = "gt", 3:ncol(.)) %>%
                                     dplyr::mutate(gt_fct = factor( gt, levels = c("0/0", "0/1", "1/1"), 
                                                                    labels = c("REF", "HET", "ALT")),
                                                                    POS = as.numeric(POS)) %>% 
                                     inner_join(metadata, ., by = "iid") %>% 
                                     dplyr::mutate(cont = rplasmodium::encode_continent(country)) %>% 
                                     dplyr::mutate(country_fct = factor(country)) 
                              
                              return(ret)
              
})


#.................................
# Make Plots
#.................................
bd$vcfplots <- purrr::map2(bd$plotdf, bd$geneid, function(x, y){
                          ret <-  x %>% 
                                  dplyr::arrange(POS) %>% 
                                  ggplot() +
                                  geom_tile(aes(x=factor(iid), y= factor(POS), fill = gt_fct)) + 
                                  facet_wrap(cont + country_fct ~ ., scales = "free_x", nrow = 1) +
                                  ggtitle(paste("Nucleotide variation for Gene", y)) +
                                  theme_classic() +
                                  theme(axis.text.x = element_blank(),
                                        plot.title = element_text(size = 12, face = "bold", hjust = 0.5))
                          return(ret)
  
})


```

```{r, results='hide'}


bd$vcfplots 



save(bd, file = "~/Documents/GitHub/VivID_Epi/WetLabWork/PvampliconDesign/pv18s.rda")

```