---
title: "Find Fasta Sequences for Primer3 based on Potential Primers"
author: "Nicholas F. Brazeau"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document:
    highlight: textmate
    theme: lumen
    toc: yes
    toc_float: yes
    toc_depth: 2
    code_folding: hide
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F, fig.width = 11, fig.height = 8)
#.................................
# imports
#.................................
library(tidyverse)
devtools::install_github("IDEELResearch/vcfRmanip")
library(vcfRmanip)
library(vcfR)
library(Biostrings)
```

# Update
Updated on April 25, 2019 to design an outer primer set (nested pcr approach) for _Pvama1_, as it showed some promise on initial run ( _Pvs47_ failed) and _pvmsp1_ is already a nested primer design. 

# Overview
Overall, uninspiring results. Will design for: 

  1. _pvs47_   
  2. _pvama1_   
  3. _pvmps1_  


```{r}

load("~/Documents/GitHub/VivID_Epi/WetLabWork/PvAmplicon/PvampliconInvestigate/potential_primer_variants.rda")

# subset to ama1 and pvs47

bd <- bd %>% 
  dplyr::filter(geneid %in% c("PVP01_1208000", "PVP01_0934200"))


# read fasta
pvp01 <- Biostrings::readDNAStringSet("~/Documents/MountPoints/mountIDEEL/resources/genomes/Pvivax/genomes/PvP01.fasta", "fasta")
chr12 <- pvp01[ pvp01@ranges@NAMES == "PvP01_12_v1" ]  # for pvs47
chr09 <- pvp01[ pvp01@ranges@NAMES == "PvP01_09_v1" ]  # for ama1
chr07 <- pvp01[ pvp01@ranges@NAMES == "PvP01_07_v1" ]  # for msp1
# pvmsp1 already designed


```

```{r}

bd$pos <- map(bd$vcfRobj, function(x){ 
  ret <- vcfR::getFIX(x)
  ret <- ret[,2]
  return(ret)
})


```

### _pvs47_ Design

From `01-Potential_Primer_interrogate.Rmd` report, I would like to capture the SNPs at the positions: `r paste("323549", "323599", "323603", "323611", "323612", "323718", "323761", "323777")`. Al together this is a 229 base pair region. However, the position 323549 is very close to the start of the gene (_starts at 323534_ which is 15 base pairs upstream). **This means I am going to have start my forward primer in a 5' UTR**. Going to have that as the lower limit and will only go 50 base-pairs up. With 20 base-pair primer expected, this will give me 

#### Conditions on `Primer-Blast`

Used the sequence below. Set the min PCR product size to 300 and the max 600. Then set organism "Plasmodium vivax (taxid:5855)". All other parameters were left at default. The "Primer Set 4" was selected as it had 55% GC content and then the lowest _Self complementarity_ score coupled with a slightly higher melt temperature than "Primer Set 3" and had a **300 bp** target (while still getting our SNPs.)

```{r}

# note these are acting as 1-based which is the same as the vcf


pvs47 <- toString(subseq(chr12, (323549-100), (323777+100)))

# forward
vmatchPattern("GACCACTAAACGCGAAGTGC", chr12)

# reverse
vmatchPattern(reverseComplement(DNAString("TTGGTGAAGCAGTTGGCTGG")), chr12)


```



### _pvama_ Design

From `01-Potential_Primer_interrogate.Rmd`, I would like to capture the SNPs at the positions: `r paste("1460429", "1460433", "1460434", "1460440", "1460441", "1460450", "1460494", "1460608", "1460628", "1460662")`. All together this is a 233 base pair region. Note, end of gene is 1460945, so have some room. 

```{r}

ama1 <- toString(subseq(chr09, (1460429-100), (1460628+100)))

vmatchPattern("TCAAATGTCCCTGCGAACCT", chr09)
vmatchPattern(reverseComplement(DNAString("TTTCCGCCCTTTTCTCTACACA")), chr09)
```

Get all the SNPs with these and don't hit variation that we have in our VCF (one site with one het call in ~800 samples) that is missing from blast. This is important if we want the primers to land. GC content only 45% but that should be OK. With all this in mind "Primer Set 1" was selected.



### _pvama_ Nested Design
Now going to design outer primers for a nested approach. Above, we know that our forward and reverse primers (eg set 1 from above) fall here, respectively: 
```{r}
prmrs <- rbind( data.frame( vmatchPattern("ATACCAAACCGACTTGCCTCA", chr09)[[1]] ),
         data.frame( vmatchPattern(reverseComplement(DNAString("TTTCCGCCCTTTTCTCTACACA")), chr09)[[1]] )
  ) %>% 
    cbind(Primer = c("Forward", "Reverse"), .)

prmrs %>% 
  DT::datatable()

```
  
Now going to leverage that information for outer primers. Going to expand out another 250 bp from the ends of the primers and ask for `Primer-blast` to find a proce with the settings: min PCR product size to 400 and the max 600. Then set organism "Plasmodium vivax (taxid:5855)". All other parameters were left at default.
```{r}

ama1.outer <- toString(subseq(chr09, 
                              (prmrs$start[prmrs$Primer == "Forward"]-250), 
                              (prmrs$end[prmrs$Primer == "Reverse"]+250)))


```

`Primer-Blast` asked to match on this *"XM_001615397.1	Plasmodium vivax apical merozoite antigen 1 (PVX_092275), partial mRNA"* .... which is a Sal1 version...which should be fine (annoying default behavior). Will check that they match and work below. The third primer set fit the bill with falling approximately 100-bp upstream and downstream of each of the inner primers designed above. Plus GC content is 55% and the reverse has a GC hammer at beginning. 


```{r}


vmatchPattern("AGTGCCTTTCTTCCAGTGGG", chr09)
vmatchPattern(reverseComplement(DNAString("GCACACGCCACCAGTTATTC")), chr09)


prmrs.all <- cbind(status = "inner", prmrs) %>% 
  rbind.data.frame(
    rbind( data.frame( vmatchPattern("AGTGCCTTTCTTCCAGTGGG", chr09)[[1]] ),
           data.frame( vmatchPattern(reverseComplement(DNAString("GCACACGCCACCAGTTATTC")), chr09)[[1]]
                       )
           ) %>% 
      cbind(status = "outer", Primer = c("Forward", "Reverse"), .)
    )

prmrs.all %>% 
  DT::datatable()


```

### _pvama_ Even Smaller Nested Design
On June 5, 2019, we decided to go after a smaller region on AMA1. I am going to nest inside the 300-bp target that I designed above. I still would like to capture the SNP at position: `r paste("1460608")`. *Going to use a heminested approach*. As above, `Primer-Blast` asked to match on this *"XM_001615397.1	Plasmodium vivax apical merozoite antigen 1 (PVX_092275), partial mRNA"* .... which is a Sal1 version...which should be fine (annoying default behavior). Found a good forward primer in primer-set 13 50% GC content and similar melting temperature (59.6 vs. 59.89).


```{r}



prmrs.all.newer <- cbind(status = "very_inner", Primer = c("Forward", "Reverse"),
                         rbind( data.frame( vmatchPattern("TCAAATGTCCCTGCGAACCT", chr09)[[1]] ),
                                data.frame( vmatchPattern(reverseComplement(DNAString("TTTCCGCCCTTTTCTCTACACA")), chr09)[[1]])))




prmrs.all.newer <- rbind.data.frame(prmrs.all, prmrs.all.newer)

prmrs.all.newer %>% 
  DT::datatable()

```
                         
                         

### _pvmsp1_ 

Where do they land?
```{r}
# inner primers published by JESS
vmatchPattern("CAAGCCTACCAAGAATTGATCCCCAA", chr07, max.mismatch = 1) 
vmatchPattern(reverseComplement(DNAString("ATTACTTTGTCGTAGTCCTCGGCGTAGTCC")), chr07)
seqinr::s2c(Biostrings::toString(chr07))[1219131:1219156]

```

### _pv18s snounou_ 

Where do they land?
```{r}
# snounou primers
snounouhitF <-vmatchPattern(reverseComplement(DNAString( "CGCTTCTAGCTTAATCCACATAACTGATAC" )), pvp01[ !grepl("Transfer", pvp01@ranges@NAMES) ] )
snounouhitR <- vmatchPattern("ACTTCCAAGCCGAAGCAAAGAAAGTCCTTA", pvp01[ !grepl("Transfer", pvp01@ranges@NAMES) ])

# apparently it is only hitting on CHR2 even though it is a multicopy gene?
vmatchPattern(reverseComplement(DNAString( "CGCTTCTAGCTTAATCCACATAACTGATAC" )), pvp01[  pvp01@ranges@NAMES == "PvP01_02_v1" ] )
vmatchPattern("ACTTCCAAGCCGAAGCAAAGAAAGTCCTTA",  pvp01[  pvp01@ranges@NAMES == "PvP01_02_v1" ] )



```



```{r}
mtdna <-  pvp01[ grepl("PvP01_MIT_v1", pvp01@ranges@NAMES) ]


vmatchPattern("ACTGTA", mtdna)


vmatchPattern("GGAGGAGGTACTGGATGGACT", mtdna)

vmatchPattern(reverseComplement(DNAString("AGGATCTCCAGCAAATGTAGGAT")), mtdna)


vmatchPattern("TTATATCCACCATTAAGTACATCACTT", mtdna)
vmatchPattern(reverseComplement(DNAString("AGTGTTAAACCTTTAGATCTTAGATG")), mtdna)




```


```{r}
# GenBank: MF197850.1
# Plasmodium simium isolate

psim <- Biostrings::readDNAStringSet("~/Downloads/sequence.fasta")

vmatchPattern("TTATATCCACCATTAAGTACATCACTT", psim)
vmatchPattern(reverseComplement(DNAString("AGTGTTAAACCTTTAGATCTTAGATG")), psim)

# GenBank: KX645965.1
# Plasmodium cf. inui
pinfui <- Biostrings::readDNAStringSet("~/Downloads/sequence (1).fasta")
vmatchPattern("TTATATCCACCATTAAGTACATCACTT", pinfui)
vmatchPattern(reverseComplement(DNAString("AGTGTTAAACCTTTAGATCTTAGATG")), pinfui)



# Can't tell the difference to psim but its a NMW isolate... not OWM?
toString(subseq(mtdna, 3726, 3880))
toString(subseq(psim, 3727, 3881))


toString(subseq(pinfui, 2985, 3880))



mtdna <- pvp01[pvp01@ranges@NAMES == "PvP01_MIT_v1"]

vmatchPattern("TTATATCCACCATTAAGTACATCACTT", mtdna)
vmatchPattern(reverseComplement(DNAString("AGTGTTAAACCTTTAGATCTTAGATG")), mtdna)


# pull in metadata
mt <- readxl::read_excel(path = "~/Documents/GitHub/VivID_Epi/WetLabWork/PvAmplicon/PvampliconInvestigate/curated_samples_20181211.xlsx") 


dv <- vcfR::read.vcfR(file = "~/Documents/MountPoints/mountedMeshnick/Sandbox/PvMtDNA_VCF/pv_combined_mtdna.raw.vcf")

mt <- readxl::read_excel(path = "~/Documents/GitHub/VivID_Epi/WetLabWork/PvAmplicon/PvampliconInvestigate/curated_samples_20181211.xlsx") 

bd <- data.frame( 
  chr = "PvP01_MIT_v1",
  start = 3691,
  end = 4214,
  strand = "+",
  name = "coxI",
  geneid = "PVX"
)




bd <- bd %>%
  dplyr::rename(seqname = chr)
bdlist <- split(bd, factor(1:nrow(bd)))

bd$vcfRobj <- purrr::map(bdlist, vcfRmanip::vcfR2SubsetChromPos, vcfRobject = dv)
bd$nvar <- unlist(purrr::map(bd$vcfRobj, function(x){ return(nrow(x@gt)) }))

#.................................
# data manip for plotting
#.................................

bd$plotdf <- purrr::map(bd$vcfRobj, function(x, metadata = mt){
                              ret <- cbind.data.frame(CHROM = vcfR::getCHROM(x),
                                     POS = vcfR::getPOS(x),
                                     vcfR::extract.gt(x, element = "GT")) %>% 
                                     gather(., key = "iid", value = "gt", 3:ncol(.)) %>%
                                     dplyr::mutate(gt_fct = factor( gt, levels = c("0/0", "0/1", "1/1"), 
                                                                    labels = c("REF", "HET", "ALT")),
                                                                    POS = as.numeric(POS)) %>% 
                                     inner_join(metadata, ., by = "iid") %>% 
                                     dplyr::mutate(cont = rplasmodium::encode_continent(country)) %>% 
                                     dplyr::mutate(country_fct = factor(country)) 
                              
                              return(ret)
              
})


#.................................
# Make Plots
#.................................
bd$vcfplots <- purrr::map2(bd$plotdf, bd$geneid, function(x, y){
                          ret <-  x %>% 
                                  dplyr::arrange(POS) %>% 
                                  ggplot() +
                                  geom_tile(aes(x=factor(iid), y= factor(POS), fill = gt_fct)) + 
                                  facet_wrap(cont + country_fct ~ ., scales = "free_x", nrow = 1) +
                                  ggtitle(paste("Nucleotide variation for Gene", y)) +
                                  theme_classic() +
                                  theme(axis.text.x = element_blank(),
                                        plot.title = element_text(size = 12, face = "bold", hjust = 0.5))
                          return(ret)
  
})



bd$vcfplots


```


